File: ios-client/Shared/Heartbeat.swift
import Foundation

final class Heartbeat {
    private var idleInterval = 0.0
    private var idleDeadline = 0.0
    private var probeDeadline = 0.0
    init() { pingSent() }
    func packetSent() { idleDeadline = ProcessInfo.processInfo.systemUptime + idleInterval }
    func pingSent() {
        let now = ProcessInfo.processInfo.systemUptime
        idleInterval = Double.random(in: 20...35)
        idleDeadline = now + idleInterval
        probeDeadline = now + Double.random(in: 50...65)
    }
    func due() -> Bool { ProcessInfo.processInfo.systemUptime >= min(idleDeadline, probeDeadline) }
}

File: sever/heartbeat.py
import secrets
import time


class Heartbeat:
    """Idle probe with jitter and a hard, jittered liveness bound during uploads."""
    def __init__(self, clock=time.monotonic, random=secrets.randbelow):
        self.clock, self.random = clock, random
        self.ping_sent()

    def packet_sent(self):
        self.idle_deadline = self.clock() + self.idle_interval

    def ping_sent(self):
        now = self.clock()
        self.idle_interval = 20 + self.random(15001) / 1000
        self.idle_deadline = now + self.idle_interval
        self.probe_deadline = now + 50 + self.random(15001) / 1000

    def due(self):
        return self.clock() >= min(self.idle_deadline, self.probe_deadline)
